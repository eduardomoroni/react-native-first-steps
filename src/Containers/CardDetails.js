/* @flow */

import React, { Component } from 'react'
import {
  View,
  StyleSheet,
  Image
} from 'react-native'
import {
  Metrics
} from '../Styles/Themes'

type CardDetailsProps = {
  card: any
}

export default class CardDetails extends Component {
  props: CardDetailsProps

  render () {
    const {
      multiverseid
    } = this.props.card
    const IMG_URL = `http://gatherer.wizards.com/Handlers/Image.ashx?multiverseid=${multiverseid}&type=card`

    return (
      <View style={styles.container}>
        <Image style={styles.image} source={{uri: IMG_URL}} />
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingTop: Metrics.navBarHeight
  },
  image: {
    flex: 1
  }
})
