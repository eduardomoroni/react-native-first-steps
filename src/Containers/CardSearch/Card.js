/* @flow */

import React, { PureComponent } from 'react'
import { View, Text } from 'react-native'
import styles from '../../Styles/CardStyle'
import { placeholdersToSymbols } from '../../Realm/Conversion/Placeholder'

// TODO: Make this component a dumb component
// TODO: Use prop-types
type CardProps = {
  card: any,
  showCardText: boolean
}

export const getLastPrinting = (printings: any) => {
  const keys = Object.keys(printings)
  const lastPrinting = printings[keys[keys.length - 1]]
  return lastPrinting.printing
}

export class Card extends PureComponent {
  props: CardProps

  render () {
    const { card, showCardText } = this.props
    const {
      name,
      text,
      manaCost,
      type,
      power,
      toughness,
      printings
    } = card
    return (
      <View style={styles.container}>
        {renderCardNameAndMana(name, placeholdersToSymbols(manaCost))}
        {renderCardTypeAndEdition(type, getLastPrinting(printings))}
        {renderTextAndPower(text, power, toughness, showCardText)}
      </View>
    )
  }
}

const renderCardNameAndMana = (leftText, rightText) => {
  return (
    <View style={styles.lineContainer}>
      <View style={styles.leftWord}>
        <Text key='cardName' style={styles.cardNameText}>{leftText}</Text>
      </View>
      <View style={styles.rightWord}>
        <Text key='cardManaCost' style={styles.mana}>{rightText}</Text>
      </View>
    </View>
  )
}

const renderCardTypeAndEdition = (leftText, rightText) => {
  return (
    <View style={styles.lineContainer}>
      <View style={styles.leftWord}>
        <Text key='cardType' style={styles.cardTypeText}>{leftText}</Text>
      </View>
      <View style={styles.rightWord}>
        <Text key='cardEdition'>{rightText}</Text>
      </View>
    </View>
  )
}

const renderTextAndPower = (text, power, toughness, showCardText) => {
  return (
    <View style={styles.cardTextContainer}>
      <View style={styles.cardText}>
        <Text key='cardText' style={styles.text}>{showCardText ? text : ''}</Text>
      </View>
      <View style={styles.cardPower}>
        <Text key='cardPowerAndToughness' style={styles.cardPowerToughness}>{power || toughness ? `${power}/${toughness}` : ''}</Text>
      </View>
    </View>
  )
}
