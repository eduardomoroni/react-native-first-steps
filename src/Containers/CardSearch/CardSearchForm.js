/* @flow */

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { View, Keyboard } from 'react-native'
import { reduxForm, Field, formValueSelector } from 'redux-form'
import { searchForCards, showFormModal } from '../../Redux/Actions'
import Styles from '../../Styles/CardSearchFormStyle'
import { valuesOf } from '../../Realm/RealmService'
import type { Dispatch } from 'redux'
import {
  DropdownInputForm,
  Modal,
  MultiSelect,
  ManaIconsBar,
  SubmitButtonForm,
  NumericInputForm,
  ModalToggle,
  TextInputForm
} from '../Components'

export class CardSearchForm extends Component {
  renderThreeFieldInRow (fieldOne, fieldTwo, fieldThree) {
    return (
      <View style={Styles.multipleFieldsPerLine}>
        <View style={Styles.leftField}>
          {fieldOne}
        </View>
        <View style={Styles.middleField}>
          {fieldTwo}
        </View>
        <View style={Styles.rightField}>
          {fieldThree}
        </View>
      </View>
    )
  }

  renderTwoFieldInRow (leftField, rightField) {
    return (
      <View style={Styles.multipleFieldsPerLine}>
        <View style={Styles.leftField}>
          {leftField}
        </View>
        <View style={Styles.rightField}>
          {rightField}
        </View>
      </View>
    )
  }

  renderModal () {
    let modalContent = {}
    let { visibleModal, cardSets, cardFormats, cardRarities } = this.props

    if (visibleModal === 'cardRarity') {
      modalContent = <Field name='cardRarity' component={MultiSelect} items={cardRarities} />
    } else if (visibleModal === 'cardSet') {
      modalContent = <Field name='cardSet' component={MultiSelect} items={cardSets} />
    } else if (visibleModal === 'cardFormat') {
      modalContent = <Field name='cardFormat' component={MultiSelect} items={cardFormats} />
    } else {
      return <View />
    }
    return (
      <Modal isVisible onModalHide={() => this.props.showModal('')} >
        {modalContent}
      </Modal>
    )
  }

  render () {
    const {
      handleSubmit,
      cardTypes,
      cardSubTypes,
      cardType,
      cardSubType,
      showModal,
      searchCards,
      cardRarity,
      cardSet,
      cardFormat
    } = this.props

    const formSubmit = (formValues) => {
      Keyboard.dismiss()
      console.log('Form sent: ', formValues)
      searchCards(formValues)
    }

    const numericOperators = ['', '<', '<=', '=', '>=', '=']

    return (
      <View style={Styles.container}>
        <View style={Styles.formContainer}>
          <Field name='cardName' component={TextInputForm} />
          <Field name='cardText' component={TextInputForm} />
          { this.renderTwoFieldInRow(
            <Field name='cardType' component={DropdownInputForm} dropdownItems={cardTypes} selectedValue={cardType} />,
            <Field name='cardSubType' component={DropdownInputForm} dropdownItems={cardSubTypes} selectedValue={cardSubType} />
        )}
          { this.renderThreeFieldInRow(
            <Field name='cardPower' component={NumericInputForm} dropdownItems={numericOperators} />,
            <Field name='cardToughness' component={NumericInputForm} dropdownItems={numericOperators} />,
            <Field name='cardCMC' component={NumericInputForm} dropdownItems={numericOperators} />
        )}
          <Field name='cardColors' component={ManaIconsBar} />
          {/* Secundary Options */}
          <Field name='cardFlavorText' component={TextInputForm} />
          { this.renderTwoFieldInRow(
            <Field name='cardArtist' component={TextInputForm} />,
            <Field name='cardCollectionNumber' component={TextInputForm} keyboardType={'numeric'} maxLength={3} />
        )}
          { this.renderThreeFieldInRow(
            <ModalToggle label='cardRarity' onPress={showModal} selected={cardRarity} />,
            <ModalToggle label='cardSet' onPress={showModal} selected={cardSet} />,
            <ModalToggle label='cardFormat' onPress={showModal} selected={cardFormat} />
        )}
          <Field name='cardColorsIdentity' component={ManaIconsBar} />
          {this.renderModal()}
        </View>
        <View style={Styles.containerFooter}>
          <SubmitButtonForm onPress={handleSubmit(formSubmit)} />
        </View>
      </View>
    )
  }
}

Field.propTypes = {
  dropdownItems: PropTypes.arrayOf(PropTypes.string),
  selectedValue: PropTypes.string
}

const mapStateToProps = (state) => {
  const selector = formValueSelector('CardSearchForm')

  const cardRarities = ['Common', 'Uncommon', 'Rare', 'Mythic Rare']
  const cardTypes = valuesOf('Type')
  const cardSubtypes = valuesOf('SubType')
  const printings = valuesOf('Printing')
  const formats = valuesOf('Legality')

  cardTypes.unshift('')
  cardSubtypes.unshift('')

  return {
    cardType: selector(state, 'cardType'),
    cardSubType: selector(state, 'cardSubType'),
    cardSubTypes: cardSubtypes,
    cardTypes: cardTypes,
    cardSets: printings,
    cardFormats: formats,
    cardRarity: selector(state, 'cardRarity'),
    cardSet: selector(state, 'cardSet'),
    cardFormat: selector(state, 'cardFormat'),
    visibleModal: state.cardSearch.visibleModal,
    cardRarities: cardRarities
  }
}

const mapDispatchToProps = (dispatch: Dispatch<*>) => {
  return {
    searchCards: (cardSearchForm) => dispatch(searchForCards(cardSearchForm)),
    showModal: (modal) => dispatch(showFormModal(modal))
  }
}

const cardSearchFormDecorated = reduxForm({form: 'CardSearchForm'})(CardSearchForm)
export default connect(mapStateToProps, mapDispatchToProps)(cardSearchFormDecorated)
