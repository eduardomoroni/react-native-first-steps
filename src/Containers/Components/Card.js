/* @flow */

import React, { Component } from 'react'
import {
  View,
  Text,
  StyleSheet
} from 'react-native'

type CardProps = {
  card: any
}

export default class Card extends Component {
  props: CardProps
  render () {
    const {
      name,
      text
    } = this.props.card
    return (
      <View style={styles.container}>
        <Text>{name}</Text>
        <Text>{text}</Text>
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1
  }
})
