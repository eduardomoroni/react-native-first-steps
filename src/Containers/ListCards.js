import React, { Component } from 'react'
import { ListView, View } from 'react-native'
import { connect } from 'react-redux'
import Card from './Components/Card'

type ListCardsProps = {
  cards: any
}

const renderRow = (card) => {
  return <Card card={{...card}} />
}

let dataSource = {}

class ListCards extends Component {
  props: ListCardsProps

  constructor (props) {
    super(props)
    const ds = new ListView.DataSource({rowHasChanged: (r1, r2) => r1 !== r2})
    dataSource = ds.cloneWithRows(props.cards)
  }

  render () {
    return (
      <View style={{flex: 1, paddingTop: 62}}>
        <ListView
          dataSource={dataSource}
          renderRow={renderRow}
        />
      </View>
    )
  }
}

const mapStateToProps = (state) => {
  const { cards } = state.cardSearch
  return {
    cards: cards
  }
}

export default connect(mapStateToProps)(ListCards)
