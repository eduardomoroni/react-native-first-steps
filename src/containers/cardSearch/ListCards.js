/* @flow */

import React, { Component } from 'react'
import { View, TouchableOpacity, ListView } from 'react-native'
// import { ListView } from 'realm/react-native'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { Actions as NavigationActions } from 'react-native-router-flux'
import styles from '../../styles/ListCardStyles'
import { Card } from './Card'
import { CardImage } from '../components'
import { cardType } from '../../types/CardType'

const showDetails = (card: cardType, cards) => {
  NavigationActions.cardSwiper({cards: cards, card: card, title: card.name})
}

let dataSource = {}
const ds = new ListView.DataSource({rowHasChanged: (r1, r2) => r1.id !== r2.id})

export class ListCards extends Component {
  constructor (props: any) {
    super(props)
    dataSource = ds.cloneWithRows(props.cards)
  }

  componentWillReceiveProps (nextProps: any) {
    if (this.props.cards !== nextProps.cards) {
      dataSource = ds.cloneWithRows(nextProps.cards)
    }
  }

  isDisplayingAsImage () {
    return this.props.showCardsAs === 'image'
  }

  renderRow = (rowData: cardType, sectionID: number, rowID: number) => {
    const { showCardText, cards } = this.props

    return (
      <TouchableOpacity
        onPress={() => showDetails(rowData, cards)}
        style={this.isDisplayingAsImage() ? styles.card : {}} >
        {this.isDisplayingAsImage() ? <CardImage card={{...rowData}} key={rowID} /> : <Card card={{...rowData}} key={rowID} showCardText={showCardText} />}
      </TouchableOpacity>
    )
  }

  render () {
    const {
      showCardsAs,
      showCardText
    } = this.props

    return (
      <ListView
        key={showCardsAs + showCardText}
        style={styles.container}
        contentContainerStyle={this.isDisplayingAsImage() ? styles.contentContainer : {}}
        renderRow={this.renderRow}
        dataSource={dataSource}
        renderSeparator={(sectionId, rowId) => <View key={rowId} style={styles.separator} />}
      />
    )
  }
}

const mapStateToProps = (state) => {
  const { cards, showCardsAs, showCardText } = state.cardSearch
  console.log('showCardText', showCardText)
  return {
    cards,
    showCardsAs,
    showCardText
  }
}

ListCards.propTypes = {
  cards: PropTypes.objectOf(cardType),
  showCardsAs: PropTypes.oneOf(['list', 'image'])
}

export default connect(mapStateToProps)(ListCards)
