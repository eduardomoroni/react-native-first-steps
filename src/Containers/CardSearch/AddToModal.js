import React, { Component } from 'react'
import { View, StyleSheet } from 'react-native'
import PropTypes from 'prop-types'
import { cardType } from '../../Types/CardType'
import { Modal, SubmitButtonForm, DropdownInputForm } from '../Components'
import WishListService from '../../Services/WishListService'
import { reduxForm, Field, formValueSelector } from 'redux-form'
import { connect } from 'react-redux'

import { RkText } from 'react-native-ui-kitten'

export class AddToModal extends Component {
  constructor () {
    super()
    this.state = {
      isVisible: true
    }
  }

  toggleModal () {
    this.state.isVisible = !this.state.isVisible
  }

  addCardAmountToList (listName, card, amount) {
    const { wishList } = this.props
    console.log(wishList[listName], listName, amount)
    WishListService.updateCardAmount(card, 2, wishList.have)
  }

  render () {
    const {
      card,
      wishList,
      howMuchCards,
      whichTradeList
    } = this.props

    const tradeLists = ['want', 'have']
    const numberRange = ['1', '2', '3', '4']

    console.log(wishList)
    return (
      <Modal isVisible onModalHide={() => this.toggleModal()} >
        <View style={styles.modal}>
          <RkText style={styles.title}>
            {`Add to TradeList`}
          </RkText>
          <Field name='whichTradeList' component={DropdownInputForm} dropdownItems={tradeLists} selectedValue={whichTradeList} />
          <Field name='howMuchCards' component={DropdownInputForm} dropdownItems={numberRange} selectedValue={howMuchCards} />
        </View>
        <View>
          <SubmitButtonForm onPress={() => this.addCardAmountToList(whichTradeList, card, howMuchCards)} />
        </View>
      </Modal>
    )
  }
}

AddToModal.propTypes = {
  card: cardType.isRequired,
  listToAdd: PropTypes.any
}

const styles = StyleSheet.create({
  container: {
    flex: 1
  },
  title: {
    fontSize: 18,
    textAlign: 'center'
  },
  modal: {
    backgroundColor: 'white',
    height: 400,
    borderTopRightRadius: 4,
    borderTopLeftRadius: 4
  }
})

const mapStateToProps = (state) => {
  const selector = formValueSelector('AddToModalForm')

  return {
    whichTradeList: selector(state, 'whichTradeList'),
    howMuchCards: selector(state, 'howMuchCards')
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
  }
}

const AddToModalDecorated = reduxForm({form: 'AddToModalForm'})(AddToModal)
export default connect(mapStateToProps, mapDispatchToProps)(AddToModalDecorated)
