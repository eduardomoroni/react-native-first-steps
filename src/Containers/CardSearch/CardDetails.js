/* @flow */

import React, { Component } from 'react'
import { View, StyleSheet } from 'react-native'
import { Metrics } from '../../Styles/Themes'
import { CardImage } from '../Components'

type CardDetailsProps = {
  card: any
}

export default class CardDetails extends Component {
  props: CardDetailsProps

  render () {
    return (
      <View style={styles.container}>
        <CardImage card={this.props.card} />
      </View>
    )
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    paddingTop: Metrics.navBarHeight
  }
})
