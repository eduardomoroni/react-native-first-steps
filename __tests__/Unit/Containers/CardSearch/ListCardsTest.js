import 'react-native'
import React from 'react'
import { shallow } from 'enzyme'
import { ListCards } from '../../../../src/Containers/CardSearch/ListCards'
import { Card } from '../../../../src/Containers/CardSearch/Card'
import { CardImage } from '../../../../src/Containers/Components'
import RealmResult from '../../../Assets/Stubs/RealmResult.json'
import styles from '../../../../src/Styles/ListCardStyles'

const props = {
  showCardsAs: 'list',
  showCardText: true,
  cards: RealmResult
}

describe('<ListCards />', () => {
  const wrapper = shallow(<ListCards {...props} />)

  it('Should Render ListCards', () => {
    expect(wrapper.prop('dataSource')._dataBlob.s1).toEqual(props.cards)
    expect(wrapper).toMatchSnapshot()
  })

  it('Should have list and image display modes', () => {
    const rowData = {dummy: 'test'}

    // renderRow as Images
    wrapper.setProps({ showCardsAs: 'image' })
    const imageRow = shallow(wrapper.instance().renderRow(rowData))
    expect(wrapper.instance().isDisplayingAsImage()).toBeTruthy()
    expect(imageRow.find(CardImage).exists()).toEqual(true)
    expect(wrapper.prop('contentContainerStyle')).toEqual(styles.contentContainer)

    // renderRow as CardAmount
    wrapper.setProps({ showCardsAs: 'list' })
    const listRow = shallow(wrapper.instance().renderRow(rowData))
    expect(wrapper.instance().isDisplayingAsImage()).toEqual(false)
    expect(listRow.find(Card).exists()).toBeTruthy()
    expect(wrapper.prop('contentContainerStyle')).toEqual({})
  })

  it('Should hide CardInfo text', () => {
    const rowData = {dummy: 'test'}

    wrapper.setProps({ ...props, showCardText: true })
    const row = shallow(wrapper.instance().renderRow(rowData))
    expect(row.find(Card).prop('showCardText')).toEqual(true)

    wrapper.setProps({ ...props, showCardText: false })
    const rowWithoutText = shallow(wrapper.instance().renderRow(rowData))
    expect(rowWithoutText.find(Card).prop('showCardText')).toEqual(false)
  })
})
