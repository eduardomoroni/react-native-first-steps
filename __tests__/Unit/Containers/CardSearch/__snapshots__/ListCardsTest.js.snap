// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ListCards /> Should Render ListCards 1`] = `
<ListViewMock
  contentContainerStyle={Object {}}
  dataSource={
    ListViewDataSource {
      "items": 0,
    }
  }
  enableEmptySections={true}
  renderRow={[Function]}
  renderScrollComponent={[Function]}
  renderSeparator={[Function]}
  style={
    Object {
      "flex": 1,
      "paddingTop": 64,
    }
  }
/>
`;
